fix: Use AWS SDK Instrument spec for common attributes

NathanielRN opened the pull request:
# Description

Update the `opentelmetry-instrumentation-aws-sdk` package to follow the [common attributes in the specification for instrumentation of the AWS SDK](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/aws-sdk.md#common-attributes), specifically, this PR only address the issues with the "common attributes".

# Updates

* `rpc.service` uses the unmodified `service.serviceIdentifier` instead of converting it to lower case
* `rpc.method` uses the unmodified `awsV2Request.operation` instead of converting it to CamelCase

Related to #160

NathanielRN commented on packages/instrumentation-aws-sdk/src/utils.ts:
```diff
@@ -28,9 +29,9 @@ export const normalizeV3Request = (
     region: string
 ): NormalizedRequest => {
-    const commandName = toCamelCase(removeSuffixFromStringIfExists(commandNameWithSuffix, 'Command'));
+    const commandName = removeSuffixFromStringIfExists(commandNameWithSuffix, 'Command');
```
In my test of a DynamoDB AWS SDK V3 command, I saw that the command was `ListTablesCommand` so we might not need the `toPascalCase()` here?

blumamir replied:
yes looks like we can omit the PascalCase here

NathanielRN commented on packages/instrumentation-aws-sdk/src/utils.ts:
```diff
@@ -28,9 +29,9 @@ export const normalizeV3Request = (
+    const commandName = removeSuffixFromStringIfExists(commandNameWithSuffix, 'Command');
     return {
-        serviceName: serviceName?.toLowerCase(),
+        serviceName,
```
In my test, whether it is `clientConfig.serviceId` or `removeSuffixFromStringIfExists(awsExecutionContext.clientName, 'Client')`, they are both already `DynamoDB`:

https://github.com/aspecto-io/opentelemetry-ext-js/blob/243e19bcebb5f95df5b4a68cbf675aa6351ca3fb/packages/instrumentation-aws-sdk/src/aws-sdk.ts#L313-L314

blumamir replied:
That is the case for services that don't have spaces in them (like DynamoDB), but it will produce different results for services like `Route 53`:
clientName is: `Route53Client` which will produce the value `"Route53"` while the `serviceId` is `"Route 53"` with a space.
That's really annoying but not sure we have any better alternative, as the the sdk is written a way which makes it really hard to extract this info correctly.
What do you think?

NathanielRN replied:
I think the spec not saying anything too detailed about this means it's up to us 🙂 I think to make it easy on ourselves, we should just stick to `api.serviceId` in V2 and `clientConfig.serviceId` like we have here in V3.

I tested this on V2 and `api.serviceId` will produce `Route 53` (with the space) so it matches with the current `clientConfig.serviceId`.

This still supports my original comment that we can delete the `removeSuffixFromStringIfExists()` call... But I'm not too familiar with the AWS JS SDKs so I think it's fine to leave it for now in case that code path gets hit 🙂 

Maybe an action item would be to add `Route 53` to the spec examples so other SDKs can follow this decision?

blumamir replied:
aws-sdk v3 is written in a very non-instrumentation-friendly fashion.
The config is only available on the `smithy-client`, but instrumentation is not possible there since the operation (or commandName) is only set later when the middleware stack is resolved. 
I tried to apply a dirty hack that stores the config on the command so it can be accessed later on when we have the command name, but it felt very shaky and non reliable, therefore I added fallback to use the data we for sure expect to be available.

I agree that we prefer to use the `clientConfig.serviceId` if possible, but code should be prepared for it to be missing. In this case, we will produce non-consistent attributes from the instrumentation.

Unfortunately, I don't think there is anything we can do about it without any refactoring of the package by the aws-sdk v3 team, but I'm always open to fresh ideas and new directions.

NathanielRN replied:
I see, thanks for telling me background on this 🙂 . So you mean that since you

> store[d] the config on the command so it can be accessed later on

it could be that the AWS SDK V3 changes it in a way that destroys it, so we should be safe and have a default in case it disappears in-between it being added and being accessed.

If I understood correctly, that makes sense to me 😄

NathanielRN commented on packages/instrumentation-aws-sdk/src/utils.ts:
```diff
@@ -14,9 +14,10 @@ export const removeSuffixFromStringIfExists = (str: string, suffixToRemove: stri
     return {
-        serviceName: service?.serviceIdentifier?.toLowerCase(),
-        commandName: toCamelCase((awsV2Request as any).operation),
+        serviceName: service?.api?.serviceId,
```
Either this or `service?.api?.className` produces the results that we expect, however `className` is a getter and this is already a string so I think this one is better.

Confirmed that it is either `S3` or `DynamoDB` with the right casing.

blumamir replied:
great, I too think this is the property we need

NathanielRN commented on packages/instrumentation-aws-sdk/src/utils.ts:
```diff
@@ -14,9 +14,10 @@ export const removeSuffixFromStringIfExists = (str: string, suffixToRemove: stri
-        serviceName: service?.serviceIdentifier?.toLowerCase(),
-        commandName: toCamelCase((awsV2Request as any).operation),
+        serviceName: service?.api?.serviceId,
+        commandName: toPascalCase(operation),
```
I couldn't find the right casing anywhere in the API, (in fact it looks like it was already in CamelCase casing everytime?) so I modified the helper to be `toPascalCase` which is otherwise known as "Upper Camel Case" or "Camel Case with the first letter uppercase".

blumamir replied:
Yeah,
Wander what is the motivation for it to be camelCase. Can see that in the [apis JSONs](https://github.com/aws/aws-sdk-js/blob/master/apis/dynamodb-2012-08-10.min.json) it's in the right PascalCase

blumamir commented on packages/instrumentation-aws-sdk/src/utils.ts:
```diff
@@ -14,9 +14,10 @@ export const removeSuffixFromStringIfExists = (str: string, suffixToRemove: stri
 
 export const normalizeV2Request = (awsV2Request: Request<any, any>): NormalizedRequest => {
     const service = (awsV2Request as any)?.service;
+    const operation = (awsV2Request as any).operation;
```
In the above line we use the `?.` optional chaining operator to protect against `awsV2Request` being nullish.
I can see that the original code was not using it, but it probably better to have it and stay consistent

NathanielRN replied:
Sounds good! I will add the operator!

blumamir reviewed (approved):
LGTM, let's discuss the open comments before merging

blumamir commented:
Did you take a look at why tests are failing?
If you need help running them or understanding what's going on let me know :)

NathanielRN commented:
> Did you take a look at why tests are failing? If you need help running them or understanding what's going on let me know :)

Thanks for your help with this! 🙂 I seem to be failing some of the SQS tests.. I can't tell why because I already updated the hook into the `ServicesExtensions.ts` file to be `SQS` not `sqs`... do you have any guesses as to what I may have missed? 😓 

```
    custom service behavior
      SQS
        ✓ sqs send add messaging attributes
        1) sqs receive add messaging attributes and context

  DynamoDB
    receive context
      ✓ should add db attributes to dynamodb request

  SQS
    receive context
      2) should set parent context in sqs receive callback
      3) should set parent context in sqs receive 'send' callback
      4) should set parent context in sqs receive promise then
      - should set parent context in sqs receive after await
      - should set parent context in sqs receive from async function
    process spans
      5) should create processing child with forEach
      6) should create processing child with map
      ✓ should not fail when mapping to non-object type
      ✓ should not fail when mapping to undefined type
      ✓ should create one processing child when throws in map
      7) should create processing child with two forEach
      ✓ should forward all parameters to forEach callback
      ✓ should create one processing child with forEach that throws
      - should create processing child with array index access
      8) should create processing child with map and forEach calls
      9) should create processing child with filter and forEach
      - should create processing child with for(msg of messages)
      - should create processing child with array.values() for loop
      - should create processing child with array.values() for loop and awaits in process
      ✓ should propagate the context of the receive call in process spans loop
```

blumamir commented:
> Thanks for your help with this! 🙂 I seem to be failing some of the SQS tests.. I can't tell why because I already updated the hook into the `ServicesExtensions.ts` file to be `SQS` not `sqs`... do you have any guesses as to what I may have missed? 😓

The service extensions also apply some logic based on the operation name. since these now have different casing, they need to be updated as well. [See here for an example](https://github.com/aspecto-io/opentelemetry-ext-js/blob/master/packages/instrumentation-aws-sdk/src/services/sqs.ts#L93)

NathanielRN commented:
@blumamir 

> The service extensions also apply some logic based on the operation name. since these now have different casing, they need to be updated as well. See here for an example

Awesome thank you this was exactly what I needed! I'm happy to see that pattern for SQS because that is exactly the pattern that PRs like #166 will benefit from 😄 

I updated those names and now all the tests all pass, but the command `yarn:test:ci:changed` just keeps looping on the tests instead of exiting. (The 6 skipped tests are marked as "pending"). Maybe I need to be using a different command?

Please run the tests again when you have a chance!

Also side note: I updated the name of V3 spans to match the name of V2 spans, since this naming is also [in the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/aws-sdk.md#common-attributes)

> The span name MUST be of the format Service.Operation as per the AWS HTTP API, e.g., DynamoDB.GetItem, S3.ListBuckets. This is equivalent to concatenating rpc.service and rpc.method with . and consistent with the naming guidelines for RPC client spans.

NathanielRN commented:
@blumamir

Thank you for helping me run the tests again! I can see now that the tests are failing specifically on `"@aws-sdk/client-sqs@3.0.0"` but they work fine for the later versions of `client-sqs` 😅.

If we still want to support this version `3.0.0`, I think the best solution is to `switch` on the `serviceName` in `normalizeV3Request` and simply map to the right value:

https://github.com/aspecto-io/opentelemetry-ext-js/blob/f9d08fb8dd95cfad7658ed979c154cf75c979262/packages/instrumentation-aws-sdk/src/utils.ts#L25-L34

i.e.

```javascript
switch(serviceName) {
    case 'sqs':
        return 'SQS';
    default:
        return serviceName;
}
```

Just as an FYI, I printed out the values of `clientConfig.serviceId` and `removeSuffixFromStringIfExists(awsExecutionContext.clientName, 'Client')` again from this line:

https://github.com/aspecto-io/opentelemetry-ext-js/blob/f9d08fb8dd95cfad7658ed979c154cf75c979262/packages/instrumentation-aws-sdk/src/aws-sdk.ts#L313-L314

and it looks like for `client-sqs@3.0.0` the values are `sqs` and `SQS` respectively. But in later versions this is fixed and they are both `SQS`. I think the precedence order is still correct as a follow up from our [previous discussion](https://github.com/aspecto-io/opentelemetry-ext-js/pull/162#discussion_r682561585) hence why I suggest the solution above.

blumamir commented:
> but the command yarn:test:ci:changed just keeps looping on the tests instead of exiting

As you probably figured out, this command runs the test on various versions of the package to make sure it works for all supported versions

> I can see now that the tests are failing specifically on "@aws-sdk/client-sqs@3.0.0" but they work fine for the later versions of client-sqs

It's hard for me to believe someone is using this old version for anything. let's just drop support for it. you can change [this file](https://github.com/aspecto-io/opentelemetry-ext-js/blob/master/packages/instrumentation-aws-sdk/.tav.yml#L14) to fix make tests run only for >=3.0.1

Thanks again for all the great work and the research you put into this :)

blumamir commented:
> Also side note: I updated the name of V3 spans to match the name of V2 spans, since this naming is also [in the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/aws-sdk.md#common-attributes)

Great, thank you :)

NathanielRN commented:
> It's hard for me to believe someone is using this old version for anything. let's just drop support for it. you can change this file to make tests run only for >=3.0.1

Thanks! I agree 🙂 I changed [.tav.yml](https://github.com/aspecto-io/opentelemetry-ext-js/blob/master/packages/instrumentation-aws-sdk/.tav.yml#L14) and the tests completed successfully on my local machine! Should be good for a re-run 🤞 

Just to confirm, this means people using `client-sqs` or `client-s3` in V3 will _still_ get tracing, they just won't have the attributes set by the `sqs.ts` `ServiceExtension` (which adds attributes based on the operation). Further, their `rpc.service` will be `sqs` not `SQS`. But I think it's find because if you are using `3.0.0` you can most likely afford to use `3.1.0` which `^3.0.0` will do anyways and is the normal as far as I'm aware for JS `package.json` dependencies 🙂 

> Thanks again for all the great work and the research you put into this :)

Absolutely! Thank you for being so helpful and quick to respond! It's made contributing fun and easy 😄

blumamir commented:
> Just to confirm, this means people using `client-sqs` or `client-s3` in V3 will _still_ get tracing, they just won't have the attributes set by the `sqs.ts` `ServiceExtension` (which adds attributes based on the operation). Further, their `rpc.service` will be `sqs` not `SQS`. But I think it's find because if you are using `3.0.0` you can most likely afford to use `3.1.0` which `^3.0.0` will do anyways and is the normal as far as I'm aware for JS `package.json` dependencies 🙂
> 
> >

Actually, since the functionality is not expected to work correctly for v3.0.0, I think it might be a good idea to not patch it as well. which means - updating the supportedVersion [here](https://github.com/aspecto-io/opentelemetry-ext-js/blob/master/packages/instrumentation-aws-sdk/src/aws-sdk.ts#L57). WDYT?

> the tests completed successfully on my local machine! Should be good for a re-run 🤞

There is still an open discussion on your issue regarding the spec and what values should be used for the rpc attributes. Do you prefer to wait few days and see if it converges? or are you waiting for this fix for something urgent?
I'm ok with both options (waiting or releasing). Let me know what you prefer.

NathanielRN commented:
> I think it might be a good idea to not patch it as well. which means - updating the supportedVersion here. WDYT?

Yeah I agree! I checked the releases and it looks like [the next release](https://github.com/aws/aws-sdk-js-v3/tags?after=v3.7.0) after `3.0.0` is `3.1.0`. `3.0.1` does not exist, so I updated that everywhere in this PR to be `3.1.0` so that the compatibility is hopefully clear.

> There is still an open discussion on your issue regarding the spec and what values should be used for the rpc attributes. Do you prefer to wait few days and see if it converges? or are you waiting for this fix for something urgent?
I'm ok with both options (waiting or releasing). Let me know what you prefer.

Thanks for being flexible while the spec evolves 🙂 I would prefer to merge this one now if that's okay with you?

I saw that [the spec has a PR to update](https://github.com/open-telemetry/opentelemetry-specification/pull/1851/files#diff-fbef7c4b2da112486e353d6088ec362441072ddf799d730091352be48f416496R26) the examples to include `Route53` explicitly instead of `Route 53`. So, in this PR I updated the README.md with the same description and added a `.replace('/s+/g', '')` on the service name so we are still up to date with them.

If the spec evolves such that there is anything else we need to change in the future hopefully it will be a smaller PR than this one and I can follow up quickly! The goal in merging early now is that it would help us towards getting consistent values across the instrumentations. 😄

Finally, I'll remove the "Fixes #160" keyword from the PR description so we can close that ourselves when when the spec PR goes in and changes don't occur for a couple of days.

blumamir commented:
merged, thanks again

blumamir commented:
released as [opentelemetry-instrumentation-aws-sdk@0.24.1](https://github.com/aspecto-io/opentelemetry-ext-js/releases/tag/opentelemetry-instrumentation-aws-sdk%400.24.1)
